src/main/java/com/emc/mongoose/storage/driver/coop/netty/http/s3/S3StorageDriver.java

```diff
@@ -157,9 +173,57 @@ public S3StorageDriver(
 		if (versioning) {
 			Loggers.MSG.info("Versioning is enabled. Make sure that if you use input items list it has version ids");
 		}
+		var authVersionTemp = storageConfig.intVal("auth-version");
+		if ((authVersionTemp != 2) && (authVersionTemp != 4)) {
+			Loggers.MSG.warn("S3 auth version is not 2 or 4. Resetting to 2.");
+			authVersionTemp = 2;
+		}
+		authVersion = authVersionTemp;
 		requestAuthTokenFunc = null; // do not use
 	}
 
+	static final class SigningKeyCreateFunctionImpl
+			implements SigningKeyCreateFunction {
+
+		String secret;
+
+		public SigningKeyCreateFunctionImpl(final String secret) {
+			this.secret = secret;
+
+		}
+
+		@Override
+		public byte[] apply(final String datestamp) {
+			return getSignatureKey(secret, datestamp);
+		}
+	}
+
+	static byte[] HmacSHA256(String data, byte[] key) throws NoSuchAlgorithmException, InvalidKeyException {
+		Mac mac = Mac.getInstance(S3Api.SIGN_V4_METHOD);
+		mac.init(new SecretKeySpec(key, "RawBytes"));
+		return mac.doFinal(data.getBytes(UTF_8));
+	}
+
+	static byte[] getSignatureKey(String key, String dateStamp) {
+		byte[] kSigning = new byte[0];
+		try {
+			byte[] kSecret = ("AWS4" + key).getBytes(UTF_8);
+			byte[] kDate = HmacSHA256(dateStamp, kSecret);
+			byte[] kRegion = HmacSHA256("", kDate); // region is left empty
```

Comment:
Could empty region cause issues while using S3 driver?

Reply:
as far as I understand some storages just take the region field that you send in auth header and use whatever you provide to do the auth procedure on their side. Amazon itself might be doing it different, but I can't check it for sure without a cluster.
Third parameter adds to cache complexity. as computeIfAbsent doesn't take several parameters, we end up with createFunctionsCache which is ugly. Using a third parameter leads to createFunctionsforcreateFunctionsCache :) I'd say leave it as is and change in case it actually fails for someone